src/livestatus.c

```diff
@@ -0,0 +1,468 @@
+  long y = strtol(s, endptr, base);
+  if (y > max) {
+    errno = ERANGE;
+    return max;
+  }
+  if (y < min) {
+    errno = ERANGE;
+    return min;
+  }
+  return y;
+} /* static long ls_strtol_subrange */
+
+// Borrowed from https://stackoverflow.com/a/29378380
+static int ls_strtoi(const char *s, char **endptr, int base) {
+#if INT_MAX == LONG_MAX && INT_MIN == LONG_MIN
+  return (int)strtol(s, endptr, base);
+#else
+  return (int)ls_strtol_subrange(s, endptr, base, INT_MIN, INT_MAX);
+#endif
+} /* static int ls_strtoi */
+
+static int ls_config(const char *key, const char *value) /* {{{ */
+{
+  if (strcasecmp("LivestatusSocketFile", key) == 0) {
+    char *socket_file = strdup(value);
+    if (socket_file == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return -1;
+    }
+
+    livestatus_obj.socket_file = socket_file;
+  }
+
+  if (strcasecmp("OnFailureMaxRetry", key) == 0) {
+    livestatus_obj.max_retry = ls_strtoi(value, NULL, 10);
+    if (errno != 0) {
+      ERROR("livestatus plugin: strtoi failed for OnFailureMaxRetry: %s",
+            STRERRNO);
+      return -1;
+    }
+  }
+
+  if (strcasecmp("OnFailureBackOffSeconds", key) == 0) {
+    livestatus_obj.backoff_sec = ls_strtoi(value, NULL, 10);
+    if (errno != 0) {
+      ERROR("livestatus plugin: strtoi failed for OnFailureBackOffSeconds: %s",
+            STRERRNO);
+      return -1;
+    }
+  }
+
+  return 0;
+} /* static int ls_config */
+
+static int unix_connect(const char *sockfile, int *sockfd) {
+  struct sockaddr_un sun = {
+      .sun_family = AF_UNIX,
+  };
+  int rc = -1;
+  int sfd = -1;
+
+  sfd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (sfd < 0) {
+    ERROR("livestatus plugin: creating client socket: %s", STRERRNO);
+    return -1;
+  }
+
+  sstrncpy(sun.sun_path, sockfile, strlen(sockfile));
+
+  do {
+    rc = connect(sfd, (struct sockaddr *)&sun, sizeof(sun));
+  } while (errno == EINTR);
+
+  if (rc < 0) {
+    ERROR("livestatus plugin: connect to socket file %s: %s", sockfile,
+          STRERRNO);
+    return -1;
+  }
+
+  *sockfd = sfd;
+
+  return 0;
+} /* unix_connect */
+
+static int ls_parse(const char *lresponse, livestatus_status_t *lstatus) {
+  int rc = -1;
+  int i = 0;
+  char **fields = NULL;
+  char *pchar = NULL;
+
+  fields = malloc(sizeof(char **) * LIVESTATUS_EXPECTED_FIELDS_RESP_NB);
+  if (fields == NULL) {
+    ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+    return -1;
+  }
+
+  for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+    fields[j] = calloc(1, sizeof(char *) * 32);
+    if (fields[j] == NULL) {
+      ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+      rc = -1;
+      goto free_all_fields;
+    }
+  }
+
+  pchar = (char *)lresponse;
+  char *wstart = (char *)lresponse;
+  int exit = 0;
+  while (!exit) {
+    if (*pchar == '\0' || *pchar == '\n') {
+      exit = 1;
+    } else if (*pchar != LIVESTATUS_FIELD_SEP) {
```

Comment:
This re-implements string tokenization, which is error prone and unnecessarily verbose. Please use *strtok_r(3)* instead.

```c
char *res = strdup(lresponse); // skip, if "lresponse" doesn't really need to be "const".
char *ptr = res;
char *saveptr = NULL;
while ((char *f = strtok_r(ptr, LIVESTATUS_FIELD_SEP, &saveptr)) != NULL) {
  ptr = NULL;
  strarray_add(&fields, &fields_num, f);
}
// use "fields"
strarray_free(fields, fields_num);
sfree(res);
```